cmd/clusterctl/cmd/generate_yaml.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/spf13/cobra"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client"
+)
+
+type generateYAMLOptions struct {
+	url           string
+	listVariables bool
+}
+
+var gyOpts = &generateYAMLOptions{}
+
+var generateYamlCmd = &cobra.Command{
+	Use:   "yaml",
+	Short: "Process yaml using clusterctl's yaml processor",
+	Long: LongDesc(`
+		Process yaml using clusterctl's yaml processor.
+
+		clusterctl ships with a simple yaml processor that performs variable
+		substitution that takes into account of default values.
+
+		Variable values are either sourced from the clusterctl config file or
+		from environment variables`),
+
+	Example: Examples(`
+		# Generates a configuration file with variable values using
+		a template from a specific URL.
+		clusterctl generate yaml --from https://github.com/foo-org/foo-repository/blob/master/cluster-template.yaml
+
+		# Generates a configuration file with variable values using
+		a template stored locally.
+		clusterctl generate yaml  --from ~/workspace/cluster-template.yaml`),
+
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return generateYAML(os.Stdout)
+	},
+}
+
+func init() {
+	// flags for the url source
+	generateYamlCmd.Flags().StringVar(&gyOpts.url, "from", "",
+		"The URL to read the template from.")
+
+	// other flags
+	generateYamlCmd.Flags().BoolVar(&gyOpts.listVariables, "list-variables", false,
+		"Returns the list of variables expected by the template instead of the template yaml")
```

Comment:
```suggestion
	generateYamlCmd.Flags().BoolVar(&gyOpts.listVariables, "variables", false,
		"Returns the list of variables expected by the template instead of the template yaml")
```

To be consistent with `envsubst`:

```
Operation mode:
  -v, --variables             output the variables occurring in SHELL-FORMAT
```

Reply:
We can do this. However, we currently have`--list-variables` flag as part of the `clusterctl config cluster` comamand. I reused this flag to be consistent between commands. 

Maybe we can keep `--list-variables` and when we do the other `generate` sub-commands we can change all the places to be consistent there as well. Thoughts?

Reply:
Let's maybe change it later, this is fine for now